### PR TITLE
Integrating trocr and torch ensembling and updating characters list refactor

### PR DESCRIPTION
**Pull request by kba**

### CODE REVIEW

File: src/eynollah/training/train.py [hunk 1]
```diff
@@ -698,6 +699,89 @@ def gen():
...
+        
+        
+        # Create a DataLoader
+        class TransformerOCRTorchDataset(TorchDataset):
+            """
+            Wraps preprocess_imgs in a format consumable by torch
+            """
+            def __init__(self, config, dir_img, dir_lab, char_to_num):
+                self.samples = list(
+                    preprocess_imgs(
```
Comment (bertsky) [inline src/eynollah/training/train.py]: Did you verify all the augmentation functions implemented for transformers/torch in `OCRDatasetYieldAugmentations` are the same in our `utils.preprocess_img` (or `utils.preprocess_img_ocr`)? Perhaps we do need a separate `utils.preprocess_img_trocr` for `task == "transformer-ocr"`?

File: src/eynollah/training/train.py [hunk 2]
```diff
@@ -698,6 +699,89 @@ def gen():
...
+                )
+
+            def __len__(self):
+                return len(self.samples)
+
+            def __getitem__(self, idx):
+                image, label = self.samples[idx]
+
+                return {
+                    "image": torch.as_tensor(image, dtype=torch.float32),
```
Comment (bertsky) [inline src/eynollah/training/train.py]: Aren't we missing `processor(image, return_tensors="pt").pixel_values.squeeze()` here?

Also, shouldn't the dict key be `"pixel_values"` instead of `"image"`?

File: src/eynollah/training/train.py [hunk 3]
```diff
@@ -698,6 +699,89 @@ def gen():
...
+                    "label": torch.as_tensor(label, dtype=torch.long),
```
Comment (bertsky) [inline src/eynollah/training/train.py]: Aren't we missing `processor.tokenizer(label, padding="max_length", max_length=max_target_length).input_ids` here?

(Followed by some `map(lambda label: label if label != processor.tokenizer.pad_token_id else -100, ...)` apparently...)

Also, shouldn't the dict key be `"labels"` instead of `"label"`?

File: src/eynollah/training/train.py [hunk 4]
```diff
@@ -698,6 +699,89 @@ def gen():
...
+            """
+            Wraps preprocess_imgs in a format consumable by torch
+            """
+            def __init__(self, config, dir_img, dir_lab, char_to_num):
+                self.samples = list(
+                    preprocess_imgs(
+                        config,
+                        dir_img,
+                        dir_lab,
+                        char_to_num=char_to_num
```
Comment (bertsky) [inline src/eynollah/training/train.py]: I don't think we have a `char_to_num` here (the equivalent of this would be `processor.tokenizer`), and `utils.preprocess_img_ocr` calling `char_to_num(tf.strings.unicode_split(...))` is not going to work for Torch, anyway...

File: src/eynollah/training/train.py [hunk 5]
```diff
@@ -698,6 +699,89 @@ def gen():
...
+                image, label = self.samples[idx]
+
+                return {
+                    "image": torch.as_tensor(image, dtype=torch.float32),
+                    "label": torch.as_tensor(label, dtype=torch.long),
+                }
+        assert characters_txt_file
+        with open(characters_txt_file, 'r') as char_txt_f:
+            characters = json.load(char_txt_f)
+        char_to_num = StringLookup(vocabulary=list(characters), mask_token=None)
```
Comment (bertsky) [inline src/eynollah/training/train.py]: That is Keras/TF-specific (for `task == "cnn-rnn-ocr"`) – see above

File: src/eynollah/training/train.py (hunk 2 above)
Comment (kba) [reply src/eynollah/training/train.py]: Possibly, both. I just refactored the code, did not intentionally introduce any changes.

File: src/eynollah/training/train.py (hunk 1 above)
Comment (kba) [reply src/eynollah/training/train.py]: I did not, good point.

They should do the same steps (only pepper_aug and pepper_bin_aug steps are switched).

But of course the label encoding and the yield values should use torch not tensorflow conventions. I'll try to find a generic way to handle that.

File: src/eynollah/training/train.py (hunk 3 above)
Comment (kba) [reply src/eynollah/training/train.py]: Yes, you're right

```python
labels = self.processor.tokenizer(txt_inp,
                                padding="max_length",
                                max_length=self.max_target_length).input_ids

labels = [label if label != self.processor.tokenizer.pad_token_id else -100 for label in labels]
```

of the previous `OCRDatasetYieldAugmentations` class must als be ported over, like the label encoding and yield values, will do.

File: src/eynollah/training/train.py (hunk 1 above)
Comment (kba) [reply src/eynollah/training/train.py]: affddd6c I added label and yield encodings to `preprocess_imgs_ocr`

File: src/eynollah/training/train.py (hunk 2 above)
Comment (kba) [reply src/eynollah/training/train.py]: 39e054e7 I am not sure where the `__getitem__` came from and why it was supposed to be doing. But from comparing with the original implementation, I think we only need the `def __iter__` which just yields all the image-label pairs now, using `pixel_values` as the key.

File: src/eynollah/training/train.py (hunk 3 above)
Comment (kba) [reply src/eynollah/training/train.py]: I moved that to `preprocess_imgs_ocr`, which encodes labels and yield values task-specifically.

File: src/eynollah/training/train.py (hunk 2 above)
Comment (bertsky) [reply src/eynollah/training/train.py]: Looking good!

Yes, seems all correct now.

File: train/requirements.txt [hunk 6]
```diff
@@ -6,8 +6,10 @@ imutils
 scipy
 tensorflow-addons # for connected_components, depublished and only compatible with tensorflow < 2.16
 tensorflow < 2.16 # for tensorflow-addons, so only needed in training
-tf_data < 2.16 # for tensorflow-addons, so only needed in training
+tf-keras < 2.16 # avoid keras 3 (also needs TF_USE_LEGACY_KERAS=1)
 protobuf < 5 # for tensorflow-addons, so only needed in training
```
Comment (bertsky) [inline train/requirements.txt:10]: Are you sure we don't need that explicit `tf-data` pin. I vaguely remember this was pulled in the wrong version somehow...

File: src/eynollah/training/train.py (hunk 4 above)
Comment (kba) [reply src/eynollah/training/train.py]: Gone as of affddd6c

File: src/eynollah/training/train.py (hunk 5 above)
Comment (kba) [reply src/eynollah/training/train.py]: Also gone in affddd6c

File: train/requirements.txt (hunk 6 above)
Comment (kba) [reply train/requirements.txt:10]: I am pretty certain that this was just a typo introduced by me in 34a9d458ce7723006d6e5ccb48045d396738d254 I meant `tf_keras` and somehow got confused with `tf.data` (which is a module of `tensorflow`). tf-data [does exist](https://pypi.org/project/tf-data) but it's an unmaintained package with 4 releases in 2018 and no GitHub repo.

Comment (bertsky) [reply train/requirements.txt:10]: ok, gotcha!
